Add support for ModuleList and Sequentials

roym899 opened the pull request:
This addresses #9. See my comment there also for some thoughts.

Basically I now store the parent hierarchy when creating the hooks and later fallback to using these if the parent has not been called already. I switched to a pre_hook, so that summary_list is in the right order. I think this simplifies the tree code.

This should work for nested modules, and independently of whether a container is called (like Sequential) or just iterated over (like ModuleList). It will not assign an id to containers which have not been called.

TylerYep commented on torchsummary/torchsummary.py:
```diff
@@ -194,26 +196,45 @@ def apply_hooks(
 ) -> None:
     """ Recursively adds hooks to all layers of the model. """
+    fallback_info = LayerInfo(module, curr_depth, None, parent_info)  # if layer is not called
+    info = LayerInfo(module, curr_depth, None, parent_info)  # only define it for type checking
```
Replace with `info = fallback_info`

TylerYep commented on torchsummary/layer_info.py:
```diff
@@ -29,7 +37,10 @@ def __init__(self, module: nn.Module, depth: int, depth_index: int):
-        return "{}: {}-{}".format(self.class_name, self.depth, self.depth_index)
+        if self.depth_index is None:
+            return "{}: {}".format(self.class_name, self.depth)
+        else:
```
Nit: Remove the else:

roym899 replied:
Fixed.

roym899 commented on torchsummary/torchsummary.py:
```diff
@@ -194,26 +196,45 @@ def apply_hooks(
         info.check_recursive(summary_list)
         summary_list.append(info)
 
+    def hook(module: nn.Module, inputs: Any, outputs: Any) -> None:
```
Fixed.

roym899 commented on torchsummary/torchsummary.py:
```diff
@@ -194,26 +196,45 @@ def apply_hooks(
 ) -> None:
     """ Recursively adds hooks to all layers of the model. """
+    fallback_info = LayerInfo(module, curr_depth, None, parent_info)  # if layer is not called
+    info = LayerInfo(module, curr_depth, None, parent_info)  # only define it for type checking
```
Fixed. Btw, it also works with only `info=LayerInfo(...)`, but I thought it'd be better to be more explicit.

roym899 commented:
I fixed the tests. For me two tests failed. 

The first was related to this change: The order of the output is changed when an exception happened, and since I think the new order is arguably more intuitive I changed the test output to match the new one.

The second only happened on my machine, where CUDA is available. The `test_pack_padded` fails, because inside `PackPaddedLSTM` the tensors created inside forward will always be created on the cpu. So I fixed it by creating them on the same device as the provided `batch`.

roym899 commented:
I've added two small test modules which use `Sequential` (by looping and by calling it) and `ModuleList` and added a normal test and an output test.
